### PR TITLE
Categorize more user errors correctly in telemetry

### PR DESCRIPTION
**Pull request by henrymercer**
Categorizing user errors correctly in telemetry helps us detect and respond to incidents faster.  This PR starts reporting two user errors as user errors in telemetry:

- Not specifying a required input for an Action
- Submitting an advanced CodeQL analysis for a repo that has default setup enabled

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

File: src/actions-util.ts [hunk 1]
```diff
@@ -24,7 +24,11 @@ const pkg = require("../package.json") as JSONSchemaForNPMPackageJsonFiles;
  * This allows us to get stronger type checking of required/optional inputs.
  */
 export const getRequiredInput = function (name: string): string {
-  return core.getInput(name, { required: true });
+  const value = core.getInput(name);
+  if (!value) {
```
Comment (adityasharad) [inline src/actions-util.ts:28]: Does this throw on a provided input with a falsy value? (I think that problem also exists currently, looking at the implementation of `getInput`.)

File: src/upload-lib.ts [hunk 2]
```diff
@@ -488,6 +489,14 @@ export async function waitForProcessing(
   }
 }
 
+function areProcessingErrorsUserError(processingErrors: string[]): boolean {
```
Comment (adityasharad) [inline src/upload-lib.ts:492]: Confusing name but hard to find a better one. How about `shouldConsiderAsUserError`?

File: src/upload-lib.ts [hunk 3]
```diff
@@ -488,6 +489,14 @@ export async function waitForProcessing(
...
+  return (
+    processingErrors.length === 1 &&
+    processingErrors[0] ===
+      "CodeQL analyses from advanced configurations cannot be processed when the default setup is enabled"
```
Comment (adityasharad) [inline src/upload-lib.ts:496]: I am a bit worried about this breaking if the upstream text changes. Do we get any richer data (eg response code) from the API response?

File: src/actions-util.ts (hunk 1 above)
Comment (henrymercer) [reply src/actions-util.ts:28]: Here's the current implementation:

https://github.com/actions/toolkit/blob/91d3933eb52b351f437151400a88ba7d57442a9b/packages/core/src/core.ts#L127

My understanding is:

- Action inputs are stored as environment variables
- `process.env[x]` is a string or undefined
- For strings, everything is truthy apart from `""`, which counts as not specifying the input

File: src/upload-lib.ts (hunk 2 above)
Comment (henrymercer) [reply src/upload-lib.ts:492]: I think that's better.  It's not clear it's about processing errors from the function name, but I think it's sufficient to provide that context in the JSDoc.

File: src/upload-lib.ts (hunk 3 above)
Comment (henrymercer) [reply src/upload-lib.ts:496]: I don't believe so — they're marked as unrecoverable delivery errors, but some unrecoverable delivery errors are user errors and others aren't.  Given the impact of this breaking is a potential monitor false alarm, I'm happy with relying on this.

Comment (adityasharad) [reply src/upload-lib.ts:496]: Good enough. Let's ask our colleagues to let us know if they change it.

**Review by adityasharad — APPROVED** [review]

**Pull request merged**